source/common/upstream/subset_lb.cc

```diff
@@ -537,14 +540,37 @@ void SubsetLoadBalancer::HostSubsetImpl::update(const HostVector& hosts_added,
   HostsPerLocalityConstSharedPtr healthy_hosts_per_locality =
       hosts_per_locality->filter([](const Host& host) { return host.healthy(); });
 
+  HostSetImpl::updateHosts(hosts, healthy_hosts, hosts_per_locality, healthy_hosts_per_locality,
+                           determineLocalityWeights(*hosts_per_locality), filtered_added,
+                           filtered_removed);
+}
+
+LocalityWeightsConstSharedPtr SubsetLoadBalancer::HostSubsetImpl::determineLocalityWeights(
+    const HostsPerLocality& hosts_per_locality) const {
   if (locality_weight_aware_) {
-    HostSetImpl::updateHosts(hosts, healthy_hosts, hosts_per_locality, healthy_hosts_per_locality,
-                             original_host_set_.localityWeights(), filtered_added,
-                             filtered_removed);
-  } else {
-    HostSetImpl::updateHosts(hosts, healthy_hosts, hosts_per_locality, healthy_hosts_per_locality,
-                             {}, filtered_added, filtered_removed);
+    if (scale_locality_weight_) {
+      const auto& original_hosts_per_locality = original_host_set_.hostsPerLocality().get();
+      const auto& original_weights = *original_host_set_.localityWeights();
+
+      auto scaled_locality_weights = std::make_shared<LocalityWeights>(original_weights.size());
+      for (uint32_t i = 0; i < original_weights.size(); ++i) {
+        // If the original locality has zero hosts, skip it. This leaves the weight at zero.
+        if (original_hosts_per_locality[i].empty()) {
+          continue;
+        }
+
+        // Otherwise, scale it proportionally to the number of hosts removed by the subset
+        // predicate.
+        (*scaled_locality_weights)[i] = (original_weights[i] * hosts_per_locality.get()[i].size()) /
```

Comment:
Do you think it would make more sense to round rather than do floor function here when we have coarse grained arithmetic, e.g. small host sets and weights?